src/eval.rs

```diff
@@ -0,0 +1,216 @@
+use crate::parse::BinOpKind;
+use crate::scope::*;
+use crate::value::*;
+use crate::EvalError;
+use gc::{Finalize, Gc, GcCell, Trace};
+use rnix::TextRange;
+use std::borrow::Borrow;
+
+type TreeResult = Result<Box<Tree>, EvalError>;
+
+/// Used to lazily calculate the value of a Tree. This should be
+/// tolerant of parsing and evaluation errors from child Trees.
+#[derive(Debug, Clone, Trace, Finalize)]
+pub enum TreeSource {
+    Literal {
+        value: NixValue,
+    },
+    Paren {
+        inner: TreeResult,
+    },
+    BinOp {
+        op: BinOpKind,
+        left: TreeResult,
+        right: TreeResult,
+    },
+    BoolAnd {
+        left: TreeResult,
+        right: TreeResult,
+    },
+    BoolOr {
+        left: TreeResult,
+        right: TreeResult,
+    },
+    Implication {
+        left: TreeResult,
+        right: TreeResult,
+    },
+    UnaryInvert {
+        value: TreeResult,
+    },
+    UnaryNegate {
+        value: TreeResult,
+    },
+}
+
+/// Syntax node that has context and can be lazily evaluated.
+#[derive(Clone, Trace, Finalize)]
+pub struct Tree {
+    #[unsafe_ignore_trace]
+    pub range: Option<TextRange>,
+    pub value: GcCell<Option<Gc<NixValue>>>,
+    pub source: TreeSource,
+    pub scope: Gc<Scope>,
+}
+
+impl std::fmt::Debug for Tree {
+    // The scope can be recursive, so we don't want to print it by default
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Tree")
+            .field("value", &self.value)
+            .field("source", &self.source)
+            .field("range", &self.range)
+            .finish()
+    }
+}
+
+impl Tree {
+    /// Lazily evaluate a Tree, caching its value
+    pub fn eval(&self) -> Result<Gc<NixValue>, EvalError> {
+        use std::ops::Deref;
+        let value_borrow = self.value.borrow();
+        if let Some(ref value) = value_borrow.deref() {
+            Ok(value.clone())
+        } else {
+            drop(value_borrow);
```

Comment:
I'm not sure I understand *why* this is needed (yes, I understand the reason for the if/else, but not for this statement). Also, it might be good to add a comment in the code :)

Reply:
Good catch! It looks like a lot of the messy borrowing code wasn't necessary https://github.com/nix-community/rnix-lsp/pull/45/commits/fd068a8eb4723d54ce4b41b5837fb3b4c7584f08